campaignion_tracking/package/src/ga4.js

```diff
@@ -0,0 +1,426 @@
+   * "cart" with only 1 slot for 1 donation.
+   *
+   * Event data: { product, currencyCode }
+   *
+   * @param {String} eventName the event name
+   * @param {object} eventData data of the event
+   * @param {object} context context of the event
+   */
+  handle_setDonationProduct (eventName, eventData, context) {
+    this.printDebug('(handle)', eventName, eventData, context)
+    this.updateContext(context)
+    if (eventData.currencyCode) {
+      this._context.donation.currencyCode = eventData.currencyCode
+    }
+    const currencyCode = this._context.donation.currencyCode || null
+    const currentProduct = this._context.donation.product || {}
+    const newProduct = eventData.product || {}
+
+    const addData = {
+      event: 'addToCart',
+      ecommerce: {
+        currencyCode: currencyCode,
+        add: {
+          products: [newProduct]
+        }
+      }
+    }
+    if (currencyCode) {
+      addData.ecommerce.currencyCode = currencyCode
+    }
+    const removeData = {
+      event: 'removeFromCart',
+      ecommerce: {
+        remove: {
+          products: [currentProduct]
+        }
+      }
+    }
+    // Only push a remove if we can assume we have pushed a valid product before.
+    const pushRemove = Object.prototype.hasOwnProperty.call(currentProduct, 'price')
+    let data = {
+      addData: addData,
+      removeData: removeData,
+      pushRemove: pushRemove
+    }
+    // Allow others to modify the data being sent to Google Analytics.
+    data = this.callChangeHook(eventName, data, this._context)
+
+    const changedNewProduct = data.addData.ecommerce.add.products[0]
+    const changedCurrentProduct = data.removeData.ecommerce.remove.products[0]
+
+    // Only change something or send an event if the donation products differ.
+    // Compare *after* any changes.
+    if (productIsEqual(changedCurrentProduct, changedNewProduct)) {
+      this.printDebug('(handle)', eventName, 'same product', eventData, context)
+      return
+    }
+
+    this._context.donation.product = changedNewProduct
+    this.saveToStorage('context', this._context)
+
+    if (data.pushRemove) {
+      this.dataLayer.push(data.removeData)
+    }
+    this.dataLayer.push(data.addData)
+  }
+
+  /**
+   * Handle "checkoutBegin".
+   *
+   * Event data: { product, currencyCode }
+   *
+   * @param {String} eventName the event name
+   * @param {object} eventData data of the event
+   * @param {object} context context of the event
+   */
+  handle_checkoutBegin (eventName, eventData, context) {
+    this.printDebug('(handle)', eventName, eventData, context)
+    this.updateContext(context)
+    const product = eventData.product || this._context.donation.product || {}
+    const currencyCode = eventData.currencyCode || this._context.donation.currencyCode || null
+    let data = {
+      event: 'checkoutBegin',
+      ecommerce: {
+        checkout: {
+          actionField: { step: 1 }, // begin == 1
```

Comment:
The comment here is funny. The one thing it says is the only thing that’s obvious from the code.

Is this something required bei GA? Is this something we define arbitrarily (and thus might need user documentation)?

Reply:
Ok, then never mind :grimacing: